sec1_intro_and_setup/test_app/Gemfile.lock

```diff
@@ -0,0 +1,218 @@
+GEM
```

Comment:
The `Gemfile.lock` file is where Bundler records the exact versions that were installed. Since developers may have different machine/ OSes, the `Gemfile.lock` file allows the developers to run the same library/project in different machines by running `bundle install`. The `bundle install` will look into the `Gemfile.lock` and install the same versions stated in this file, rather than installing the most recent versions.

Reply:
Why need it to be the same exact version?

Reply:
Running different versions on different machines could lead to broken tests, etc. Hence, one must never directly edit the lock file.

Reply:
Broken tests so?

Reply:
Can you elaborate more about the question?

Reply:
You mentioned, running different version on different machines could lead to broken tests. 
My question is, whats the consequences of having broken tests? If broken, then just fix it?

This is cause, if the test is broken (or something else), the following is one main cause I can think of:
Test & codes are for V1 but the one broken might be running a version that is not 1. If is lesser than 1, just upgrade. If higher than 1, fix the test. then the one at lower version just upgrade

Since its going to be fixed anyway, why need to have exact version?

Reply:
Having the exact version makes repositories more friendly towards new and existing contributors. Ideally, anyone should be able to clone the repo, run `bundle install`, and have passing tests. If you don't check in your `Gemfile.lock`, new contributors can get different versions of your dependencies, and run into failing tests that they don't know how to fix.

Plus, it would be troublesome for you to manually resolve all of the dependencies in your application, including dependencies of dependencies. And even once you do that successfully, you would need to redo that work if you wanted to update a particular gem. For instance, if you wanted to update the `rails` gem, you would need to find all of the gems that depended on dependencies of Rails, and find new versions that satisfy the new versions of Rails' requirements.  If you made a mistake in the manual dependency resolution process, you would not get any feedback about conflicts between different dependencies, resulting in subtle runtime errors.

Reply:
```
run into failing tests that they don't know how to fix
```

Then remove the test then all will be okay already isn't it?

Reply:
Well, you can remove the test but that does not solve the problem. How do you ensure that there will be no failing tests in the future even if it does not run into failing tests after you remove the test this time? If you want to make changes to `Gemfile`, the safer way will be getting the exact same version first, then run `bundle install`, Bundler will update only the gems that you modified.

Reply:
Well, you can don't add any tests. Then no issue isn't it?

Reply:
Then how do you know the development environment is set up and working correctly?

Reply:
huh?
YOu don't need test to set up your development environment.
Like right now, your this project don't have any tests

Reply:
Why don't need a test?
Let say you are contributing to a gem developed by others, and you want to fix a bug but you get a different version of dependencies. Wouldn't getting the same version of dependencies be easier for you to find the bug? If getting a different version of dependencies, how do you ensure the bug is still the same bug you want to fix given that you don't add any test?

Reply:
Whether we should or shouldn't write test is a different conversation altogether.
Right now, the question is, if you do not write test, the application still works. Your answer is all about fixing whats in the test. What if an application that is used by thousands of users has no test written at all. Then your answer is irrelevant. Thus, your answer still doesn't asnwer the question

Reply:
So the question is `Why need exact same version?` ?

Reply:
In a way yes.

Reply:
If you specify the exact version of every gem, then you’ll never have unexpected breaks while upgrading.

Reply:
Er, if you don't upgrade, then no problem.
*You don't need to upgrade your system for it to work.
http://falive.org/ is running on angularjs 1 with libraries that are no longer available. but it still works.

Reply:
The code might not be compatible. For example, dev1 starts a new app with a Rails gem version `~>6.0`, a few months later, dev1 shares the repo without Gemfile.lock with dev2 to collaborate on the app. Dev2 runs `bundle install` and get version 6.1, but is not compatible with some of the code dev1 wrote that only works with version 6.0.

Reply:
Then, why not put in gemfile the exact version you are using? Then no matter what, it will always use the same version and no need this file

Reply:
The Gemfile.lock records the versions last time your code successfully worked. Specifying the exact version in Gemfile would not guarantee the same because gems usually declare a range of versions for their dependencies.

Reply:
Whats stopping me from specifying the exact version in Gemfile?

Reply:
If you specify the exact version in Gemfile, the dependencies might not specify the exact version (the same goes to sub-dependecies). If you specify the exact version of dependencies manually, you would likely run into conflicts very easily with other gems that use the same sub-dependency.

Reply:
That is incorrect. If you specify the exact version in Gemfile, it will pull the exact version for the sub dependency.

Reply:
> Whats stopping me from specifying the exact version in Gemfile?

Reply:
>Take into consideration the risks of unexpected breaks due to inconsistent gems during deployment, it might be better to specify the major version instead of the exact version as it reduces the risks of unexpected breaks.

If you say about unexpected breaks, I'd rather use exact version. Then I know it will definitely work. IF you specify major version only, you have a chance it will break if the gem maintainer didn't do things properly. So, what are you talking about here?

>Gemfile contains all the gems for your use and Gemfile.lock also contains the dependencies needed by the gems. (from here, we know Gemfile.lock has more gems than Gemfile)
ok

Reply:
> If you say about unexpected breaks, I'd rather use exact version. Then I know it will definitely work. IF you specify major version only, you have a chance it will break if the gem maintainer didn't do things properly. So, what are you talking about here?

If you are very confident that your application won't have any unexpected breaks, then use exact versions. If you are not sure, specify the major version, although there is still a possibility that there will have unexpected breaks, the possibility is lower than when you specify the exact version.

Reply:
What I meant is when you set the exact version of your gem in your Gemfile, the other developers who are using your gem will get the exact version you specified in your Gemfile but not including the dependencies version since you didn't include the Gemfile.lock (considering Gemfile take over the job of Gemfile.lock) and the dependencies version varies because Gemfile only got names and version range but did not have dependencies. 

Let say you specified the exact version in Gemfile and during that time your dependencies version works fine with your gem, and after a few months, the dependencies versions updated and some of the features got deleted in the dependency's version, then your gem will no longer be usable, and that causes unexpected breaks. That's why when you are 100% sure even the dependencies updated with the Gemfile specify the exact version (no Gemfile.lock), and your gem will still working fine, go ahead specify the exact version in Gemfile.

Reply:
And what I said previously was that if you specified a version in your Gemfile, it will be the same version for the dependency too.

```
the dependencies version varies because Gemfile only got names and version range 
```
You are contradicting your self here. You said `range` here yet `you set the exact version of your gem`.
**Exact** vs **Range**.
How can you compare apple to oranges and assume your logic make sense?

Exact vs Exact. Thats what I am saying.

eg:
If in gemfile you say: lib1: "1.0.0", it will load dep1: 1.0.0
Everyone that run bundle install will get lib1(1.0.0) & its dep1(1.0.0).

It won't be:
lib1: 1.0.0
dep1: not 1.0.0 magically.

---

So, again, re read the thread and pay attention to the details of what you are trying to explain. So far you have only been confusing me and yourself.

Reply:
> Why don't we just set exact version in Gemfile
1. Setting the exact version is the job of Gemfile.lock.

> What is stopping us from setting exact version in Gemfile
1. Setting an exact version in the Gemfile stops you from doing `bundle update gem_that_you_want_to_update` whereas `~>` or no version allows you to run `bundle update gem_that_you_want_to_update` and get the latest (minor) version.

Reply:
> Setting the exact version is the job of Gemfile.lock.

Who says? I can set it in Gemfile

> Setting an exact version in the Gemfile stops you from doing bundle update gem_that_you_want_to_update whereas ~> or no version allows you to run bundle update gem_that_you_want_to_update and get the latest (minor) version.

If I set the exact version in Gemfile, I don't want to upgrade it cause of possible issues. So, this is not a concern

Reply:
> What is stopping us from setting exact versions in Gemfile

Then nothing more is stopping you from specifying the exact version. As the system grows larger with more and more dependencies, try not to make the specifications of the dependencies too tight, else version lock might occur, but also not too loose because of version promiscuity.

Reply:
ok. So, how should we define it then?
lets use the example of the gem: `lib1`. How will you write it in your gemfile (assuming its `1.x` now)

Reply:
If the gem cannot function without the other gems on which it depends., then use `=`, eg: `gem 'lib1', '= 1.0.0'`
If the gem can function without the other modules on which it depends, use `~>`, e.g: `gem lib1, ~> 1.0` (allows new functions and improvements of the gem minor version)

Reply:
How would you know if the gem can function or not function with the dependency?

---

If you tell me to check its dependency, think about it.

Gem1 depends on Gem2
Gem3 depends on Gem4
Gem2 depends on Gem4
Gem4 depends on Gem5
Gem6 depends on Gem2
etc etc

It will be crazy to assume you will and can go through all the dependencies especially if it start going down multiple levels.

Reply:
Since you cannot know for sure which dependencies functions with the gem, make the gems as soft dependencies according to the semantic versioning scheme in the beginning (most of the time use `~>`). Use `=` when you are monkey-patching a library outside of its public API or manually pin the gem version when break happens due to update of multiple dependencies at once.

Reply:
Not correct. But, lets move on. You'll understand once you start working on this